File: Stage27/Java/Problem1707.java
package Stage27.Java;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class Problem1707 {
    static ArrayList<Integer>[] graph;
    static int[] color;
    static boolean isBipartite;

    static void bfs(int start) {
        Queue<Integer> queue = new LinkedList<>();
        queue.offer(start);
        color[start] = 1;

        while (!queue.isEmpty()) {
            int cur = queue.poll();

            for (int next : graph[cur]) {
                if (color[next] == 0) {
                    color[next] = -color[cur];
                    queue.offer(next);
                } 
                else if (color[next] == color[cur]) {
                    isBipartite = false;
                    return;
                }
            }
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();

        int K = Integer.parseInt(br.readLine());

        while (K-- > 0) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            int V = Integer.parseInt(st.nextToken());
            int E = Integer.parseInt(st.nextToken());

            graph = new ArrayList[V + 1];
            color = new int[V + 1];
            isBipartite = true;

            for (int i = 1; i <= V; i++) {
                graph[i] = new ArrayList<>();
            }

            for (int i = 0; i < E; i++) {
                st = new StringTokenizer(br.readLine());
                int u = Integer.parseInt(st.nextToken());
                int v = Integer.parseInt(st.nextToken());

                graph[u].add(v);
                graph[v].add(u);
            }

            for (int i = 1; i <= V; i++) {
                if (color[i] == 0) {
                    bfs(i);
                }
                if (!isBipartite) break;
            }

            sb.append(isBipartite ? "YES\n" : "NO\n");
        }

        System.out.print(sb);
    }
}

File: Stage27/Python/Problem1707.py
import sys
from collections import deque

input = sys.stdin.readline

def bfs(start):
    queue = deque([start])
    color[start] = 1

    while queue:
        cur = queue.popleft()

        for nxt in graph[cur]:
            if color[nxt] == 0:
                color[nxt] = -color[cur]
                queue.append(nxt)
            elif color[nxt] == color[cur]:
                return False
    return True

def main():
    K = int(input())

    for _ in range(K):
        V, E = map(int, input().split())

        global graph, color
        graph = [[] for _ in range(V + 1)]
        color = [0] * (V + 1)

        for _ in range(E):
            u, v = map(int, input().split())
            graph[u].append(v)
            graph[v].append(u)

        is_bipartite = True

        for i in range(1, V + 1):
            if color[i] == 0:
                if not bfs(i):
                    is_bipartite = False
                    break

        print("YES" if is_bipartite else "NO")

if __name__ == "__main__":
    main()